onnx/defs/shape_inference.h

```diff
@@ -354,6 +354,56 @@ inline void propagateElemTypeFromSequenceInputToOutput(InferenceContext& ctx, si
   }
 }
 
+inline void propagateElemTypeFromOptionalInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex) {
```

Comment:
I think this is more specialized/complex than it needs to be. I think it is better if the utility function works for arbitrary optional types. Further, the extra checks on the output-type is unnecessary. (I realize that this comes from the template used for other types, but it would be useful to start cleaning this up for the new code at least.) It should basically be something like "*outputType = inputType" (or as required by the protobuf generated classes to copy/clone).

Reply:
I think this include file is becoming too large. It would be better to have just the interface/signature in the shape_inference.h file and move implementation to shape_inference.cc

Reply:
I refactored and deleted some part of the code. I think the code looks simpler now.